File: library/src/main/java/org/apache/apex/malhar/lib/window/Window.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.apex.malhar.lib.window;

import org.apache.hadoop.classification.InterfaceStability;

/**
 * This interface describes the classes that represent individual windows.
 *
 * @param <WINDOW> window type the object of this class can call compareTo
 *
 * @since 3.5.0
 */
@InterfaceStability.Evolving
public interface Window<WINDOW extends Comparable<WINDOW>> extends Comparable<WINDOW>
{
  long getBeginTimestamp();

  long getDurationMillis();

  /**
   * Global window means there is only one window that spans the entire life time of the application
   */
  class GlobalWindow implements Window<GlobalWindow>
  {

    /**
     * The singleton global window
     */
    public static final GlobalWindow INSTANCE = new GlobalWindow();

    private GlobalWindow()
    {
    }

    @Override
    public long getBeginTimestamp()
    {
      return 0;
    }

    public long getDurationMillis()
    {
      return Long.MAX_VALUE;
    }

    @Override
    public boolean equals(Object other)
    {
      return this == other;
    }

    @Override
    public int compareTo(GlobalWindow o)
    {
      return 0;
    }

    @Override
    public String toString()
    {
      return "[GlobalWindow]";
    }
  }

  /**
   * TimeWindow is a window that represents a time slice
   *
   * @param <WINDOW> window type the object of this class can call compareTo
   */
  class TimeWindow<WINDOW extends TimeWindow<WINDOW>> implements Window<WINDOW>
  {
    protected final long beginTimestamp;
    protected final long durationMillis;

    private TimeWindow()
    {
      // for kryo
      this.beginTimestamp = -1;
      this.durationMillis = 0;
    }

    public TimeWindow(long beginTimestamp, long durationMillis)
    {
      this.beginTimestamp = beginTimestamp;
      this.durationMillis = durationMillis;
    }

    /**
     * Gets the beginning timestamp of this window
     *
     * @return the begin timestamp
     */
    @Override
    public long getBeginTimestamp()
    {
      return beginTimestamp;
    }

    /**
     * Gets the duration millis of this window
     *
     * @return the duration
     */
    @Override
    public long getDurationMillis()
    {
      return durationMillis;
    }

    @Override
    public boolean equals(Object other)
    {
      if (other instanceof TimeWindow) {
        TimeWindow otherWindow = (TimeWindow)other;
        return this.beginTimestamp == otherWindow.beginTimestamp && this.durationMillis == otherWindow.durationMillis;
      } else {
        return false;
      }
    }

    @Override
    public int hashCode()
    {
      int result = (int)(beginTimestamp ^ (beginTimestamp >>> 32));
      result = 31 * result + (int)(durationMillis ^ (durationMillis >>> 32));
      return result;
    }

    @Override
    public int compareTo(TimeWindow o)
    {
      long diff = (this.getBeginTimestamp() + this.getDurationMillis()) - (o.getBeginTimestamp() + o.getDurationMillis());
      if (diff == 0) {
        return Long.signum(this.getBeginTimestamp() - o.getBeginTimestamp());
      } else {
        return Long.signum(diff);
      }
    }

    @Override
    public String toString()
    {
      return "[TimeWindow " + getBeginTimestamp() + "(" + getDurationMillis() + ")]";
    }
  }

  /**
   * SessionWindow is a window that represents a time slice for a key, with the time slice being variable length.
   *
   * @param <K> the key type for the session
   */
  class SessionWindow<K> extends TimeWindow<SessionWindow<K>>
  {
    private final K key;

    private SessionWindow()
    {
      // for kryo
      this.key = null;
    }

    public SessionWindow(K key, long beginTimestamp, long duration)
    {
      super(beginTimestamp, duration);
      this.key = key;
    }

    public K getKey()
    {
      return key;
    }

    @Override
    public boolean equals(Object other)
    {
      if (!super.equals(other)) {
        return false;
      }
      if (other instanceof SessionWindow) {
        @SuppressWarnings("unchecked")
        SessionWindow<K> otherSessionWindow = (SessionWindow<K>)other;
        if (key == null) {
          return otherSessionWindow.key == null;
        } else {
          return key.equals(otherSessionWindow.key);
        }
      } else {
        return false;
      }
    }

    @Override
    public int hashCode()
    {
      return (key.hashCode() << 16) | super.hashCode();
    }

    @Override
    public int compareTo(SessionWindow<K> o)
    {
      int val = super.compareTo(o);
      if (val == 0) {
        if (this.getKey() instanceof Comparable) {
          return ((Comparable<K>)this.getKey()).compareTo(o.getKey());
        } else {
          return Long.compare(this.getKey().hashCode(), o.getKey().hashCode());
        }
      } else {
        return val;
      }
    }

    @Override
    public String toString()
    {
      return "[SessionWindow key=" + getKey() + " " + getBeginTimestamp() + "(" + getDurationMillis() + ")]";
    }
  }
}

File: python/apex-python/src/pyapex/functions/window.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

class TriggerType:
  EARLY = 1
  ON_TIME = 2
  LATE = 3
  @staticmethod
  def isValid(trigger_type):
    if trigger_type < TriggerType.EARLY:
      raise Exception("Incorrect Trigger Type")
    if trigger_type > TriggerType.LATE:
      raise Exception("Incorrect Trigger Type")
    return True

class AccumulationMode:
  DISCARDING = 1
  ACCUMULATING = 2
  ACCUMULATING_AND_RETRACTING = 3
  @staticmethod
  def isValid(type):
    if type < TriggerType.DISCARDING:
      raise Exception("Incorrect Accumulation Mode")
    if type > TriggerType.ACCUMULATING_AND_RETRACTING:
      raise Exception("Incorrect Accumulation Mode")
    return True

class TriggerOption(object):
  triggers =  []
  accumulation_mode = None
  firingOnlyUpdatedPanes = False
  @staticmethod
  def at_watermark():
    triggerOption = TriggerOption()
    trigger = Trigger(TriggerType.ON_TIME)
    triggerOption.triggers.append(trigger)
    return triggerOption

  def withEarlyFiringsAtEvery(self,*args,**kwargs):
    trigger = None
    if 'count' in kwargs:
      trigger = CountTrigger(TriggerType.EARLY, kwargs['count'])
    if 'duration' in kwargs:
      trigger = TimeTrigger(TriggerType.EARLY, kwargs['duration'])
    if trigger is None:
      raise Exception("Unsufficent for trigger")
    self.triggers.append(trigger)
    return self

  def withLateFiringsAtEvery( self, *args, **kwargs ):
    trigger = None
    if 'count' in kwargs:
      trigger = CountTrigger(TriggerType.LATE, kwargs['count'])
    if 'duration' in kwargs:
      trigger = TimeTrigger(TriggerType.LATE, kwargs['duration'])
    if trigger is None:
      raise Exception("Unsufficent for trigger")
    self.triggers.append(trigger)
    return self

  def discardingFiredPanes(self):
    self.accumulation_mode = AccumulationMode.DISCARDING
    return self

  def accumulatingFiredPanes(self):
    self.accumulation_mode = AccumulationMode.ACCUMULATING
    return self

  def accumulatingAndRetractingFiredPanes(self):
    self.accumulation_mode = AccumulationMode.ACCUMULATING_AND_RETRACTING
    return self

  def firingOnlyUpdatedPanes(self):
    self.firingOnlyUpdatedPanes = True
    return self

  @staticmethod
  def get_java_trigger_options(trigger_option, gateway):
    _jtrigger_option = None
    for trigger in trigger_option.triggers:
        if trigger.trigger_type == TriggerType.ON_TIME:
          _jtrigger_option = gateway.jvm.TriggerOption.AtWatermark()
        elif trigger.trigger_type == TriggerType.EARLY:
          if isinstance(trigger, TimeTrigger):
            _jduration = gateway.jvm.Duration(trigger.duration)
            _jtrigger_option = _jtrigger_option.withEarlyFiringsAtEvery(_jduration)
          else:
            _jcount = gateway.jvm.Duration(trigger.count)
            _jtrigger_option = _jtrigger_option.withEarlyFiringsAtEvery(_jcount)
        elif trigger.trigger_type == TriggerType.LATE:
          if isinstance(trigger, TimeTrigger):
            _jduration = gateway.jvm.Duration(trigger.duration)
            _jtrigger_option = _jtrigger_option.withLateFiringsAtEvery(_jduration)
          else:
            _jcount = gateway.jvm.Duration(trigger.count)
            _jtrigger_option = _jtrigger_option.withLateFiringsAtEvery(_jcount)
    return _jtrigger_option

class Trigger(object):
  trigger_type = None

  def __init__(self,trigger_type):
    self.trigger_type = trigger_type

class TimeTrigger(Trigger):
  duration = None

  def __init__(self, trigger_type, duration):
    super(TimeTrigger,self).__init__(trigger_type)
    self.duration = duration

class CountTrigger(Trigger):
  count = None

  def __init__(self, trigger_type, count):
    super(CountTrigger, self).__init__(trigger_type)
    self.count = count





